mrrc/storage.py

```diff
@@ -0,0 +1,235 @@
+            if not os.path.isfile(full_file_path):
+                logger.warning('Warning: file %s does not exist during uploading. Product: %s',
+                               full_file_path, product)
+                return
+            logger.info('Uploading %s to bucket %s', full_file_path, bucket_name)
+            fileObject = bucket.Object(path)
+            existed = self.__file_exists(fileObject)
+            sha1 = read_sha1(full_file_path)
+            if not existed:
+                f_meta = {}
+                if sha1.strip() != "":
+                    f_meta[CHECKSUM_META_KEY] = sha1
+                if product:
+                    f_meta[PRODUCT_META_KEY] = product
+                if len(f_meta) > 0:
+                    fileObject.put(Body=open(full_file_path, 'rb'), Metadata=f_meta)
+                else:
+                    fileObject.upload_file(full_file_path)
+            else:
+                logger.info('File %s already exists, check if need to update product.',
+                            full_file_path)
+                f_meta = fileObject.metadata
+                checksum = f_meta[CHECKSUM_META_KEY] if CHECKSUM_META_KEY in f_meta else ""
+                if checksum != "" and checksum.strip() != sha1:
+                    logger.error('Error: checksum check failed. The file %s is different from the '
+                                 'one in S3. Product: %s', path, product)
+                    return
+
+                prods = []
+                try:
+                    prods = f_meta[PRODUCT_META_KEY].split(",")
+                except KeyError:
+                    pass
+                if product not in prods:
+                    logger.info('File %s has new product, updating the product %s', full_file_path,
+                                product)
+                    prods.append(product)
+                    self.__update_file_metadata(fileObject, bucket_name, path,
+                                                {PRODUCT_META_KEY: ",".join(prods)})
+
+            logger.info('Uploaded %s to bucket %s', full_file_path, bucket_name)
+            return
+
+        self.__do_path_cut_and(
+            file_paths=file_paths,
+            fn=path_upload_handler,
+            root=root)
+
+    def upload_metadatas(self, meta_file_paths: List[str], bucket_name: str, product: str,
+                         root="/"):
+        """ Upload a list of metadata files to s3 bucket. This function is very similar to
+        upload_files, except:
+            * The metadata files will always be overwritten for each uploading
+
+            * The metadata files' checksum will also be overwritten each time
+        """
+        bucket = self.__get_bucket(bucket_name)
+
+        def path_upload_handler(full_file_path: str, path: str):
+            if not os.path.isfile(full_file_path):
+                logger.warning('Warning: file %s does not exist during uploading. Product: %s',
```

Comment:
There is a \#TODO comment here before, please do not remove it as we need it as a reminder to fix the missing logic later

Reply:
I removed it because it was causing a linting issue. Can we add it as a requirement for one of the ongoing tasks so that it's completed before release? My thinking is that if we add a TODO in code then more likely than not we'll end up just leaving it there.